### PR TITLE
Refactoring: disentangle brax training/architecture, single-file

### PR DESCRIPTION
**Pull request by SimpleGeometry**
Clarity changes
- Merged losses.py and networks.py into train.py
- Removed much brax entanglement, especially typehints which were more opaque than helpful + structures which are difficult to understand and edit since it refers back to brax repo (evaluator/replay buffer/make_policy for policy inference are still there)
- Removed a lot of clarity-hurting multilining

Non-clarity changes
- Networks changed to be MLPs with residual blocks, motivated by Kevin's findings. The default architecture is still the same: 256-width, 2 hidden layer, which is expressed as 1 residual block of depth 2. To have a regular MLP, just use 1 residual block. Currently there is no flag to specify the residual block size; it's set to 2 by default.
- Rendering every test epoch (i.e. once per num_eval). This takes a decent amount of time especially with >10 evals, so we may want to include an option to render less frequently.

Tested on Ant and appears to be at parity. However, this refactor introduces a lot of changes and it's possible there are still issues.

===============

The docs/comments are still WIP, though it's mergeable with dev:
- ~~Comments in train.py with suggestions/questions/thoughts, which probably should be removed before pushing to main branch (above make_policy, above compute_energy, under the docstring in train, in sgd_step, above additional_sgds, and under train_epoch_with_timing).~~ Comments are in [dce9e61](https://github.com/MichalBortkiewicz/JaxGCRL/pull/26/commits/dce9e61fe65d3e0dd6f237e206b977dc7e05b998) but not in final commit.
- Many docstrings are outdated (arguments removed and added).

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -1,118 +1,317 @@
...
+    critic_params = {"sa_encoder": sa_encoder_params, "g_encoder": g_encoder_params}
+    critic_state = TrainState.create(apply_fn=None, params=critic_params, tx=optax.adam(learning_rate=critic_lr))
+
+    # Put everything together into TrainingState
+    training_state = TrainingState(env_steps=jnp.zeros(()), gradient_steps=jnp.zeros(()), 
+                                   actor_state=actor_state, critic_state=critic_state, alpha_state=alpha_state)
+    training_state = jax.device_put_replicated(training_state, jax.local_devices()[:num_local_devices_to_use])
+    return training_state
+
+# Problem: we have made this a single file, but the abstraction is bad. Users will have a difficult time understanding the
```
[the remark this answers is not in the document]
Comment (MichalBortkiewicz) [reply src/train.py:115]: Later perhaps we can encapsulate some logic into `Agent` class (including `energy`, `log_softmax`  and maybe loss functions? )

**Review by MichalBortkiewicz — APPROVED** [review]
Looks good, and works as previous implementations (tested with ant and ant_u_maze). I see +- how we can reduce the cofnitive load by encapsulating some agent logic into new agent class and there is some room for improvement regarding SGD logic refactoring but it is for later PR

**Comment by MichalBortkiewicz** [comment]
Also, let's remove long comments about code structure and potential refactoring before merging to master

File: src/train.py [hunk 2]
```diff
@@ -1,118 +1,317 @@
...
+    """
+    MLP with residual connections: residual blocks have $block_size layers. Uses swish activation, optionally uses layernorm.
+    """
+    output_size: int
+    width: int = 1024
+    num_blocks: int = 1
+    block_size: int = 2
+    use_ln: bool = True
+    @nn.compact
+    def __call__(self, sa):
```
[the remark this answers is not in the document]
Comment (SimpleGeometry) [reply src/train.py:38]: Thanks for the catch, I was adapting this from the SAEncoder where it was called sa. Maybe just call it `x` with standard convention?

File: src/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SimpleGeometry) [reply src/train.py:115]: Noted, I like trying to separate it out but want to make sure the abstraction is reasonably tight. If we call it Agent, while this file is Train, is it clear to users which functionality should be in which file, and does it make it easier to understand rather than needing to learn what functionality is where? For example, I wonder about loss and energy functions; we could put energy functions in Agent but loss feels like it should go in Train, but then separating out loss and energy might be weird.

No clear ideas on this front from me, just noting some thoughts.

File: src/train.py [hunk 3]
```diff
@@ -297,332 +493,186 @@ def train(
...
+        return jax.lax.scan(f, (training_state, env_state, buffer_state, key), (), length=num_prefill_actor_steps)[0]
+    
     prefill_replay_buffer = jax.pmap(prefill_replay_buffer, axis_name=_PMAP_AXIS_NAME)
 
-    def additional_sgds(
-        training_state: TrainingState,
-        buffer_state: ReplayBufferState,
-        key: PRNGKey,
-    ) -> Tuple[TrainingState, ReplayBufferState, Metrics]:
+    # Should we rename the below functions? Additional_sgds seems to be "train enough for one replay buffer sample", scan_additional_sgds
```
[the remark this answers is not in the document]
Comment (SimpleGeometry) [reply src/train.py:619]: Agreed on doing in later PRs.

File: src/train.py [hunk 4]
```diff
@@ -272,15 +465,18 @@ def train(
         None
 
     """
-
+    # Main goal for clarifying this section: distill down to "bookkeeping" and "collect/train/eval loop", where
+    # it's easy to tell the high-level structure of the collect/train, so people can quickly trace a relevant
+    # component to swap or understand. E.g. replace losses, replace training, replace architectures, how to
+    # interact with the environment (and what the function call exactly looks like).
+    
+    # This mostly seems like bookkeeping, wonder if we could find a clean summary of what bookkeeping is happening?
```
[the remark this answers is not in the document]
Comment (SimpleGeometry) [reply src/train.py:473]: Added.

**Comment by SimpleGeometry** [comment]
> Also, let's remove long comments about code structure and potential refactoring before merging to master

All the aforementioned comments have been removed. For posterity, comments can be found in the "Merge with dev branch" commit, [dce9e61](https://github.com/MichalBortkiewicz/JaxGCRL/pull/26/commits/dce9e61fe65d3e0dd6f237e206b977dc7e05b998).

File: utils.py [hunk 5]
```diff
@@ -279,13 +279,13 @@ def print_times(self):
         print(f"time to jit: {self.times[1] - self.times[0]}")
         print(f"time to train: {self.times[-1] - self.times[1]}")
         
-    def progress(self, num_steps, metrics):
+    def progress(self, num_steps, metrics, make_policy, params, env, exp_dir, exp_name):
```
[the remark this answers is not in the document]
Comment (SimpleGeometry) [reply utils.py:282]: Done, and argument added as "visualization frequency".
